File: 0045-jump-game-ii/0045-jump-game-ii.py
class Solution(object):
    def jump(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        n = len(nums)
        if n == 1:
            return 0
        
        jumps = 0
        curr_end = 0
        furthest = 0
        
        for i in range(n - 1):
            furthest = max(furthest, i + nums[i])
            if i == curr_end:
                jumps += 1
                curr_end = furthest
                
        return jumps

File: 0045-jump-game-ii/0045-jump-game-ii.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var jump = function(nums) {
    if (nums.length === 1) {
        return 0;
    }

    let jumps = 0;
    let end = 0;
    let maxReach = 0;

    for (let i = 0; i < nums.length - 1; i++) {
        maxReach = Math.max(maxReach, i + nums[i]);
        if (i === end) {
            jumps++;
            end = maxReach;
        }
    }

    return jumps;
};
